src/vsr/journal.zig

```diff
@@ -703,83 +843,505 @@ pub fn Journal(comptime Replica: type, comptime Storage: type) type {
+        ///   match checksum           _   _   _   _   _   _   _  !1   0   0   0   1
+        ///   match op                 _   _   _   _   _   _   _  !1   <   >   1  !1
+        ///   match view               _   _   _   _   _   _   _  !1   _   _  !0  !1
+        ///   decision (replicas>1)  vsr vsr vsr vsr vsr vsr vsr nil fix vsr vsr eql
+        ///   decision (replicas=1)                  fix fix
+        ///
+        /// Legend:
+        ///
+        ///    0  false
+        ///    1  true
+        ///   !0  assert false
+        ///   !1  assert true
+        ///    _  ignore
+        ///    <  header.op < prepare.op
+        ///    >  header.op > prepare.op
+        ///
+        /// A "valid" header/prepare:
+        /// 1. has a correct checksum, AND
+        /// 2. one of the following:
+        ///   * command=reserved
+        ///   * command=prepare AND correct cluster AND is in its proper slot (op % slot_count).
+        ///
+        ///
+        /// Regarding cases @E and @F for *single-replica clusters*:
+        ///
+        /// Most likely the redundant header write was corrupted/lost (respectively).
+        /// Since single-replica clusters never rewrite ops, it can be locally repaired by copying
+        /// over the prepare's header.
+        ///
+        /// However, another (astronomically low) possibility is this:
+        ///
+        /// 1. A prepare is written to slot 12.
+        /// 2. A redundant header is written to slot 12.
+        /// 3. Continue normal operation; wrap around the WAL.
+        /// 4. A prepare is written to slot 12, but lost.
+        /// 5. A redundant header is written to slot 12.
+        /// 6. Replica crash.
+        /// 7. On recovery, the old prepare (from step 1) is read and used to "repair" the corrupt
+        ///    redundant header.
+        ///
+        /// This would be a safety violation — we assume that the likelihood of this event is
+        /// negligible.
+        fn recover_prepares_on_read(completion: *Storage.Read) void {
+            const read = @fieldParentPtr(Self.Read, "completion", completion);
+            const self = read.self;
+            const replica = @fieldParentPtr(Replica, "journal", self);
+            assert(!self.recovered);
+            assert(self.recovering);
+            assert(read.destination_replica == null);
+
+            const slot = Slot{ .index = @intCast(u64, read.checksum) };
+            assert(slot.index < slot_count);
+
+            // `recover_headers_on_read` either sets both dirty+faulty, or neither.
+            if (self.dirty.bit(slot)) {
+                assert(self.faulty.bit(slot));
+            }
+
+            const Decision = enum {
+                /// The header and prepare are identical; no repair necessary.
+                eql,
+                /// Reserved; clear dirty/faulty, no repair necessary.
+                nil,
+                /// If replica_count>1: Repair with VSR `request_prepare`. Mark dirty, clear faulty.
```

Comment:
> Is it because here the header is valid, but from a previous log cycle? We have the prepare valid on disk. We haven't lost the latest prepare. Therefore, we don't need to set the prepare as faulty. But we do want to repair the header on disk.

Yes, that's correct. I'll add some docs to clarify. Though when `replica_count=1` it will also `fix` in cases `@E` and `@F`, and the former has an invalid header.

We set dirty so that the redundant header is repaired. We don't set faulty because we assume that we have the valid message.

> If my understanding of the above is correct, why can't we do a local repair, i.e. use the prepare's header to write the header without a network trip?

Yes, it's largely to simplify by using existing `request_prepare` machinery. Though actually repairing redundant headers is a bit trickier to do correctly than it might seem, since it may accidentally convert a neighboring case `@D` to `@H`.

Haven't thought this through, but maybe when writing a sector of redundant headers, any of those headers that are faulty are deliberately written as junk, so that if the replica crashes/recovers those entries will still be faulty and not reserved.

Reply:
Implemented in https://github.com/coilhq/tigerbeetle/pull/91/commits/1cc8ccf09ccdca4e92f1b710be427a86ff72ccaa.